lib/internal/modules/cjs/loader.js

```diff
@@ -218,6 +218,22 @@ function tryExtensions(p, exts, isMain) {
   return false;
 }
 
+// find the longest (possibly multi-dot) extension registered in
+// Module._extensions
+function findExtension(filename) {
+  const name = path.basename(filename);
```

Comment:
If you're only interested in the extension, is calling `path.basename()` necessary? You could just scan backwards for dots, checking that there are no path separators in between:
```js
let dot = -1;
let sep = -1;

for (let i = filename.length; --i > 0; /* empty */) {
  const c = filename[i];
  if (c === '.')
    dot = i;
  else if (c === '/') { // FIXME handle backslash on windows
    sep = i;
    break;
  }
}

if (dot > sep + 1) {
  const ext = filename.slice(dot);
  if (Module._extensions[ext]) return ext;
}

return '.js';  // dot file or no extension
```
Only one slice == less garbage to collect.

Reply:
The current algorithm scans forwards to short-circuit at the longest matched extension. So say you have a file like `mr.robot.coffee.md`, and a loader registered for `.coffee.md`. The current algorithm will iterate like:
- Is `.robot.coffee.md` registered? No, continue.
- Is `.coffee.md` registered? Yes, break.

This way, even if `.md` is _also_ registered, the `.coffee.md` loader takes precedence. This allows multi-dot extensions like `.es6.js`, for example.

If I’m reading your code correctly, it’s trying to find the longest possible multi-dot extension, which for this example would be `.robot.coffee.md`, a false match. We want multi-dot extensions to work without prohibiting periods elsewhere in the filename.

Reply:
Perhaps naming the function `findRegisteredExtension` would be more self-descriptive?

Reply:
@richardlau Yes, or perhaps `findLongestRegisteredExtension` would be even more descriptive (because we’re returning `.coffee.md` instead of `.md`, if both of those happened to be registered).